text/0000-inherent-trait-impl.md

```diff
@@ -0,0 +1,152 @@
+
+The feature is implemented using a new attribute which can be applied to trait
+`impl` blocks:
+
+```rust
+pub trait Bar {
+    fn bar(&self);
+}
+
+pub trait ExtBar: Bar {
+    fn ext_bar(&self);
+}
+
+impl<T: Bar> ExtBar for T {
+    fn ext_bar(&self) { }
+}
+
+pub struct Foo;
+
+#[inherent]
+impl Bar for Foo {
+    fn bar(&self) { println!("foo::bar"); }
+}
+
+// works thanks to specialization
+#[inherent]
+impl ExtBar for Foo { }
+
+impl Foo {
+    fn foo(&self) { println!("foo::foo"); }
+}
+```
+
+The methods `bar` and `ext_bar` are now callable on any instance of `Foo`,
+regardless of whether the `Bar` and `ExtBar` traits are currently in scope,
+or even whether the `Bar` trait is publically visible. In other words if `Bar`
+and `ExtBar` defined in one crate and `Foo` and another, user of `Foo` will be
+able to explicitly depend only on the crate which defines `Foo` and still use
+the inherent traits methods.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `inherent` attribute in the above example makes the `impl` block equivalent to:
+
+```rust
+impl Foo {
+    #[inline]
+    pub fn bar(&self) { <Self as Bar>::bar(self); }
+
+    fn foo(&self) { println!("foo::foo"); }
+}
+```
+
+Any questions regarding coherence, visibility or syntax can be resolved by
+comparing against this expansion, although the feature need not be implemented
+as an expansion within the compiler.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Increased complexity of the language.
+- Hides use of traits from users.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+- Define inherent traits either on a type `T` or on an `impl T { .. }` block.
+- Implement as part of the delegation RFC.
+- Do nothing: users may choose to workaround the issue by manually performing the expansion if required.
+
+The most viable alternative is delegation proposal, although arguably inherent
+traits and delegation solve different problems with the similar end result.
+The former allows to use trait methods without importing traits and the latter
+to delegate methods to the selected field. Nethertheless delegation RFC and
+this RFC can be composable with each other:
+
+```Rust
+struct Foo1;
+struct Foo2;
+
+#[inherent]
+impl T1 for Foo1 {
+    fn a() {}
+}
+
+#[inherent]
+impl T2 for Foo2 {
+    fn b() {}
+}
+
+struct Bar {
+    f1: Foo1,
+    f2: Foo2,
+}
+
+impl Bar {
+    // all methods from `T1` will be delegated as well
+    // though `T1` will not be implemented for `Bar`
+    delegate * to f1;
+}
+
+// method `b` will be accessable on `Bar` without importing `T2`
+#[inherent]
+impl T2 for Bar {
+    delegate * to f2;
+}
+```
+
+# Prior art
+[prior-art]: #prior-art
+
+- https://github.com/rust-lang/rfcs/pull/2309 (previous closed PR)
+- https://github.com/rust-lang/rfcs/issues/1880
+- https://github.com/rust-lang/rfcs/issues/1971
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
```

Comment:
Would something like having a `pub use Trait::*` in the type's `impl` block work syntax-wise?

```rust
trait Trait { ... }

struct T;

impl Trait for T { ... }

impl T {
    pub use Trait::*; // We expose the methods of the trait here
}

Reply:
It probably could, but personally I don't like it for several reasons:
1) If type does not have any inherent methods, it would require to add an `impl` block only for "imports". `#[inherent]` is shorter, local to the applied trait, and clearly a separate language feature from usual imports.
2) Semantics are somewhat confusing and surprising. What `use Trait::*;` without `pub` would mean? Currently `use` is used to bring items to the current scope, extending it to "`impl` scopes" looks like quite a stretch to me.
3) I don't think that capability to expose only some methods from a trait will be needed often in practice, so almost all imports will use `Trait::*`.

Reply:
I see what you mean...

Reply:
We'd discussed many conflicting uses for `use` in `impl` blocks, with delegation being a particularly poor choice.  Actually `impl T { pub use Trait; }` makes way more sense than most semantically, although still too strange I think.  

We might consider `pub(visibility) impl Trait for T { .. }` if we wanted to make inherent traits really standard practice and avoid attributes.